Add SavePolarizationNXcanSAS Algorithm

adriazalvarez opened the pull request:
### Description of work

This PR introduces a new algorithm to save polarized data for SANS in the NXcanSAS format based on the open proposal : https://wiki.cansas.org/index.php?title=NXcanSAS_v1.1

This PR follows the recommendations of the proposal and adds the mandatory parameters to the save algorithm, which are the minimum set necessary for a proper characterization of the experimental data and metadata. 

To implement the algorithm, I have extended works previously done on NXcanSAS to derive the SaveNXPolarizedNXcanSAS algorithm from a SaveNXcanSASBase algorithm which is the class that 
manages also SaveNXcanSAS. This class is organized in methods for saving data for either polarized and non-polarized workspaces. 
Most functions used for writing the information to the file in the nx format are defined in the helper file NXcanSASHelper as well as NXcanSASDefinitions for the definitions of the NXcanSAS groups, attributes and units. 

The main different of this algorithm with respect to the original NXcanSAS file savers is the dimensionality of the stored signal data on the file.  Previously the sans signal was either 1D (row) or a matrix to represent 2D data for either SANS or SANS2D, respectively. Polarized data increases the number of dimension by adding the polarization state of the sample as attributes of the signal axes, and thus, a hypermatrix containing the signal for all polarization states on the experiment is recorded on the file. 

Fixes #38505 <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

- I have tried to organized the methods present on the NXcanSASHelper file and reduce the number of lines, and at the same time combining with methods to save polarized data, thus a lot of refactoring has been neccessary
- I have changed the spin state validator to add an optional 3rd spin state (as this algorithm accepts 3 spin states (+1,0,-1) in contrast to others.  This change is quite non-invasive and doesn't alter the way in which spin state validator oeprates and it is constructed.
- The test suite for SaveNXcanSASTest has been also refactored and reestructured, by promoting many helper methods onto a base class that can be derived and used in other test like SavePolarizedNXcanSASTest. 
- I have added an instrument definition file on `/instrument/unit-testing/` folder for a fake polarized sans instrument to be able to simplify testing of some of the polarized metadata and not rely on ISIS instruments. 


### To test:
You can call the algorithm with different variations using this script: 

```

def createPolGroup(size, is2d):
    num_banks = 2 if is2d else 1
    workspaces = []

    for n in range(size):
        name=f'out_{n}'
        ws = CreateSampleWorkspace( OutputWorkspace=f'out_{n}', Function='User Defined', UserDefinedFunction=f'name=Lorentzian, Amplitude=1000, PeakCentre={n+1}, FWHM=1', 
                                    XUnit='MomentumTransfer', NumBanks=num_banks, BankPixelWidth=1, XMin=0, XMax=16.5, BinWidth=0.1)
        if is2d:
            ConvertSpectrumAxis(ws, OutputWorkspace=f'out_{n}', Target = "ElasticQ", EMode = "Direct", EFixed =1.)
        LoadInstrument(f'out_{n}',InstrumentName='ZOOM', RewriteSpectraMap=False)
        AddSampleLog(f'out_{n}', 'mag_field','2',LogType = 'Number', LogUnit = 'G')
        workspaces.append(name)
       
    group = GroupWorkspaces(workspaces)
    
createPolGroup(4, False)
SavePolarizedNXcanSAS("group",'test1D_full', InputSpinStates="+1+1,-1+1,+1-1,-1-1")


createPolGroup(4, True)
SavePolarizedNXcanSAS("group",'test2D_full', InputSpinStates="+1+1,-1+1,+1-1,-1-1")


createPolGroup(2, False)
SavePolarizedNXcanSAS("group",'test1D_half', InputSpinStates="+10,-10")


#testing also the metadata here
createPolGroup(2, True)
SavePolarizedNXcanSAS("group",'test2D_half_metadata', InputSpinStates="0+1,0-1",
MagneticFieldStrengthLogName="mag_field" , PolarizerComponentName="short-polarizer",
AnalyzerComponentName="helium-analyzer",
FlipperComponentNames="RF-flipper",
MagneticFieldDirection="1,2,3")

```

- Additionally, please build docs and check documentation is properly rendered and readable. 
   Beware, test code for the SavePolarizedNXcanSAS docs better be changed when we have the LoadPolarizedNXcanSAS algorithm available. 


<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

adriazalvarez committed:
Change spin state validator to input std::string and accept an optional extra spin paramter

adriazalvarez committed:
Refactor constructor for spin state validator where necessary

adriazalvarez committed:
add save pol algorithm

adriazalvarez committed:
Add cansas and nexus definitions for polarization

adriazalvarez committed:
Add polarization to SaveNXcanSAS base

adriazalvarez committed:
Refactor NXcanSASHelper file to save polarized metadata and data

adriazalvarez committed:
Add missing spin combinations

adriazalvarez committed:
Add helper struct for handling spin vectors and add error signal to file

adriazalvarez committed:
Add default suffix for sas entry

adriazalvarez committed:
Add a struct to handle spin vectors more gracefully, and reorganied code better

adriazalvarez committed:
Add some validation for input spin strings and reorganized redundant code

adriazalvarez committed:
Save docs file with basic info

adriazalvarez committed:
typo in definitions

adriazalvarez committed:
Remove private class methods and inherit from nxcansasfiletest class to access those methods

adriazalvarez committed:
Add helper test class for SaveNXcanSASTest and SavePolarizedNXcanSASTest

adriazalvarez committed:
Add release notes

adriazalvarez committed:
Clean SavePolarized class

adriazalvarez committed:
ammend release note typo

adriazalvarez committed:
Add nxcansas polarization states to polarization correction helpers

adriazalvarez committed:
Add SavePolarized test suite

adriazalvarez committed:
Refactor NXcanSAS test helpers to accomodate polarized data

adriazalvarez committed:
Update definitions

adriazalvarez committed:
Refactor, reorganized and clean NXcanSAS helper library

adriazalvarez committed:
- Refactor, reorganized and clean base algorithm
- Improve validation for polarized data
- Add definitions for standard properties
- Improve polarized component map creation

adriazalvarez committed:
Modify signature of prepareFilename after refactor of helper file

adriazalvarez committed:
Modify signature of prepareFilename after refactor of helper file in SavePol algorithm

adriazalvarez committed:
Refactor Load and Save NXcanSAS test after changing the structure of transmission parameters in helper test file

adriazalvarez committed:
Refactor, reorganized and clean helper test class
Add more private methods for testing polarized data

adriazalvarez committed:
Flesh out docs

adriazalvarez committed:
Add instrument file to unit_testing folder in instruments with a polarized sans instrument

adriazalvarez committed:
Fix some cppcheck warnings

adriazalvarez committed:
Changes after rebase

adriazalvarez committed:
Fix doxygen typo and include error

adriazalvarez committed:
Try to fix docs typo

MialLewis reviewed (changes requested):
Thanks for all the work Adrian, the code is of a high quality and the output files look good.

A few minor points, the main one that needs actioning is the moving of the spin state validator to the kernel. I don't think we can justify adding the algorithm library as a dependency of the data handling library.

I wonder if it makes sense to do that in a separate PR then rebase this one on that?

There are also a few long functions which I think would benefit from splitting out into smaller functions. You've often added comments where I feel it'd be natural to split them off. After a few hours of reviewing the long functions are a bit of a headache!

adriazalvarez committed:
Refactor return of extra pairs

adriazalvarez committed:
Reword error message and use output stream to simplify conditional error message

adriazalvarez committed:
- Bring spin state validator from kernel
- Remove dependencies with Mantid::algorithms
- Homogeneize all const to west const

adriazalvarez committed:
Remove link to mantid algorithms from data handling

adriazalvarez committed:
Add better error message in spin state validator

adriazalvarez committed:
Add separator parameter to split string into vector

adriazalvarez committed:
Update file date

adriazalvarez commented on Framework/DataHandling/inc/MantidDataHandling/NXcanSASDefinitions.h:
```diff
@@ -112,6 +116,42 @@ const std::string sasBeamAndSampleSizeUnitAttrValue = "mm";
+  const std::string &getComponentType() const { return m_type; }
+
+private:
+  std::string m_type;
```
I'm not sure about this one. I'd prefer to keep like this, but we could ask SANS scientist.

adriazalvarez commented on Framework/DataHandling/inc/MantidDataHandling/NXcanSASDefinitions.h:
```diff
@@ -112,6 +116,42 @@ const std::string sasBeamAndSampleSizeUnitAttrValue = "mm";
+const std::string sasDataPolarizationUnitAttr = "none";
+
+struct InstrumentPolarizer {
+  explicit InstrumentPolarizer(const std::string &type, const std::string &name) : m_type(type), m_name(name) {};
```
You are right that in this case there's no danger of implicit construction , it's kind of out of habit.

adriazalvarez commented on Framework/Algorithms/test/PolarizationCorrections/SpinStateValidatorTest.h:
```diff
@@ -59,7 +59,7 @@ class SpinStateValidatorTest : public CxxTest::TestSuite {
   }
 
   void testSpinStateDuplicateEntry() {
-    auto validator = std::make_shared<SpinStateValidator>(std::unordered_set<int>{2, 3}, false, false);
```
Yes I think so, it was probably unintended as 0 is the default value anyway.

adriazalvarez commented on Framework/Algorithms/inc/MantidAlgorithms/PolarizationCorrections/SpinStateValidator.h:
```diff
@@ -42,5 +43,7 @@ class MANTID_ALGORITHMS_DLL SpinStateValidator : public Kernel::TypedValidator<s
   const std::string m_para;
```
With the nxcansas format, some spin states have two characters ("-1","+1")

adriazalvarez commented on Framework/DataHandling/src/SaveNXcanSASBase.cpp:
```diff
@@ -180,55 +326,52 @@ std::map<std::string, std::string> SaveNXcanSASBase::validateStandardInputs() {
 
-  const auto scaledBgSubWorkspace = getPropertyValue("BackgroundSubtractionWorkspace");
-  const auto scaledBgSubScaleFactor = getPropertyValue("BackgroundSubtractionScaleFactor");
+  addProcess(sasEntry, workspace, transmissionCan);
```
I think there is not.

MialLewis replied:
You removed the check around `addProcess`, is this not needed if the transmission can run is not guarenteed?
```
  if (transmissionCan) {
    addProcess(sasEntry, workspace, transmissionCan);
  } else {
    addProcess(sasEntry, workspace);
  }
```

adriazalvarez commented on Framework/DataHandling/src/NXcanSASHelper.cpp:
```diff
@@ -314,173 +472,235 @@ void addDetectors(H5::Group &group, const Mantid::API::MatrixWorkspace_sptr &wor
+
+  auto const instrumentAttr = InstrumentPolarizer(componentType, componentName);
+  auto instrumentGroup =
+      group.exists(sasInstrumentGroupName)
```
Yes, the group should exists at this point of the algorithm. I'll just open the group here.

adriazalvarez commented on Framework/DataHandling/test/SavePolarizedNXcanSASTest.h:
```diff
@@ -0,0 +1,229 @@
+    assertSavedFileFormat(savePolAlg);
+  }
+
+  void test_algorithm_saves_with_no_issue_for_test_data_half_polarized_1D_and_2D() {
```
In my computer it takes 4.75 sec. Not the fastest on data handling, but also there are few far slower tests. I've been here following the style of other nxcansas load/save tests, but I agree with you the approach here can be though of as a system test.

adriazalvarez committed:
Move SaveNXcanSAS and SavePolarizedNXcanSAS property names to nxcansasdefinitions

adriazalvarez committed:
- Add helper class for storing data dimensions information
- Split save polarized fucntion into smaller functions

adriazalvarez committed:
- Remove property names from cansasbase
- Split validator of polarized properties into smaller validator functions

adriazalvarez committed:
Use initilization list for points and histogram variables

MialLewis reviewed (dismissed):
Thanks your response to my comments.

I've added a few more - I think it makes sense to respond to these at the same time as the gatekeeper review as this is such a large PR.

@cailafinn would you mind flagging any variables that should be `const` as you gatekeep?

cailafinn commented on Framework/DataHandling/src/SaveNXcanSASBase.cpp:
```diff
@@ -84,87 +96,192 @@ void SaveNXcanSASBase::initStandardProperties() {
+  std::map<std::string, std::string> result;
+  const Workspace_sptr workspace = getProperty(StandardProperties::INPUT_WORKSPACE);
+
+  if (!workspace->isGroup())
+    result.emplace(StandardProperties::INPUT_WORKSPACE,
+                   "Input Workspaces for polarized data can only be workspace groups.");
+  else {
```
Rather than a big if/else, it may be worth simply returning here instantly as we know we can't do any more checks. 
```suggestion
  if (!workspace->isGroup()) {
    result.emplace(StandardProperties::INPUT_WORKSPACE,
                   "Input Workspaces for polarized data can only be workspace groups.");
    return result;
  }

```

cailafinn commented on Framework/DataHandling/src/SaveNXcanSASBase.cpp:
```diff
@@ -180,55 +297,52 @@ std::map<std::string, std::string> SaveNXcanSASBase::validateStandardInputs() {
-    addProcess(sasEntry, workspace, transmissionCan);
-  } else {
-    addProcess(sasEntry, workspace);
-  }
```
Follow up from Mial's question about removing the existance check fo the transmission run:
https://github.com/mantidproject/mantid/pull/39118/files#r2059736841

adriazalvarez replied:
I removed the overloaded `addProcess` function and check if there is a transmission can inside the function. Looks cleaner to me this way.

cailafinn commented on Framework/DataHandling/src/SaveNXcanSASBase.cpp:
```diff
@@ -6,37 +6,49 @@
+  return (unit && unit->unitID() == unitToCompareWith);
+}
+
+void areAxesNumeric(const MatrixWorkspace_sptr &workspace, int numberOfDims = 1) {
```
Can `numberOfDims` be `const`?

cailafinn commented on Framework/DataHandling/src/SaveNXcanSASBase.cpp:
```diff
@@ -293,22 +447,61 @@ void SaveNXcanSASBase::addData(H5::Group &group, const Mantid::API::MatrixWorksp
+ * @param wsGroup: Group workspace containing 1D or 2D SANS polarized workspaces
+ * @param path: File system path in which to store the data
+ */
+void SaveNXcanSASBase::savePolarizedGroup(const WorkspaceGroup_sptr &wsGroup, const std::filesystem::path &path) const {
```
Just trying to get my head around the file setup, but this seems like it and some of the other polarisation-specific methods are only relevant to the `PolarizedNXCanSAS` algorithm. Is there a reason it's in the base class rather than the subclass?

adriazalvarez replied:
I was trying to keep all the file management in the base and the cansas helper. Would you prefer them to be in the algorithms ?

MialLewis replied:
I agree with you on both counts, it feels like `addPolarizedData` should not be in the base class. Let's however get this in for now and raise an issue to refactor this shortly.

If we have time during the release period I'm sure we can qualify it as maintenance...

cailafinn commented on Framework/DataHandling/test/NXcanSASFileTest.h:
```diff
@@ -0,0 +1,577 @@
+
+  void do_assert_detector(const H5::Group &instrument, const std::vector<std::string> &detectors) {
+    for (auto &detector : detectors) {
+      std::string detectorName = sasInstrumentDetectorGroupName + detector;
```
There's a lot of variables in this file that seem like they could be `const`?

cailafinn commented on Framework/Kernel/src/SpinStateValidator.cpp:
```diff
@@ -40,15 +45,17 @@ std::string SpinStateValidator::checkValidity(const std::string &input) const {
-                     m_anti + "."
-               : "The spin states must consist of two characters, either a " + m_para + " or a " + m_anti + ".";
+    std::ostringstream msg;
+    msg << "The format for the spin states is invalid, every comma separated value should contain ";
+    msg << (m_acceptSingleStates ? "either one or two spin states " : "two spin states ") << "from the set " << m_para
+        << "," << m_anti << (!m_extra.empty() ? "," + m_extra : "") << ".";
+
```
Thanks for improving this!

cailafinn commented on Framework/Kernel/src/VectorHelper.cpp:
```diff
@@ -548,17 +548,16 @@ bool isConstantValue(const std::vector<double> &arra) {
  * a vector of doubles.
  * @param listString :: a string like "0.0 1.2" or "2.4, 5.67, 88"
- * @return a vector of doubles
+ * @param separator :: Specify the string in which to split into vector. Default space and comma: " ,"
```
We should be clear here that it's using `is_any_of`. I.e. Any of the characters used in the string will be used as delimiters.

`"0,1 2, 3"` would still produce a `[0, 1, 2, 3]` vector with the `separator` set to ` ,`. 

Maybe renaming `separator` to `separators` would be more clear? 

It would also be good to add a test for this behaviour.

cailafinn commented on docs/source/release/v6.13.0/SANS/New_features/38505.rst:
```diff
@@ -0,0 +1 @@
+- Add new :ref:`algm-SavePolarizedNXcanSAS` algorithm to saved reduce polarized SANS data in NXcanSAS format.
```
Typo?
```suggestion
- Add new :ref:`algm-SavePolarizedNXcanSAS` algorithm to save reduced polarized SANS data in NXcanSAS format.
```

cailafinn reviewed (changes requested):
These are mostly minor things and a couple of things I had questions about. Code looks great on the whole though, nice work!

adriazalvarez commented on Framework/DataHandling/src/NXcanSASHelper.cpp:
```diff
@@ -5,42 +5,212 @@
+
+using AttrMap = std::map<std::string, std::string>;
+
+std::optional<size_t> findWorkspaceIndexForSpinState(const std::vector<std::string> &spinStates,
+                                                     const std::string &targetState) {
+  size_t index = std::ranges::find(spinStates, targetState) - spinStates.cbegin();
+  return index == spinStates.size() ? std::nullopt : std::optional(index);
+}
```
#39252

adriazalvarez committed:
Add consts where appropriate in NXcanSASFileTest

adriazalvarez committed:
Fix grammar typo in release notes

adriazalvarez committed:
- Rename parameter separator to separators
- Specify the role of separators splitting multiple characters
- Add test for multiple character separators

adriazalvarez committed:
- Add consts where appropriate
- Rename Pin, Pout to pIn, pOut
- Prefix getters in DataDimensions struct with get
- Fix various docstring typos

adriazalvarez committed:
- Improve error handling in areAxisNumeric function
- Fix various typos in docstrings
- Add consts were appropriate

adriazalvarez committed:
Use std iterator function to check non numeric axis and avoid cppcheck warning

cailafinn reviewed (approved):
This looks good to me now. Thanks for all your work on this!
